lib/SPIRV/SPIRVTypeScavenger.cpp

```diff
@@ -53,156 +130,498 @@
 
 using namespace llvm;
 
+namespace {
+static bool isTypeVariable(Type *T, unsigned &TypeVarNum) {
+  if (auto *TET = dyn_cast<TargetExtType>(T))
+    if (TET->getName() == "typevar") {
+      TypeVarNum = TET->getIntParameter(0);
+      return true;
+    }
+  return false;
+}
+
+/// Convert Ty to a type that can be unified with a type-variable-ified L, given
+/// that either or both types may have an indirection.
+/// For example, adjust(L, true, Ty, false) will extract the element type of Ty
+/// for unifying with L.
+/// In this method, L is expected to be a type of a value, while Ty (and the
+/// return value) will use TypedPointerType instead of PointerType.
+static Type *adjustIndirect(Type *L, bool LIndirect, Type *Ty, bool TIndirect) {
+  if (LIndirect)
+    Ty = cast<TypedPointerType>(Ty->getScalarType())->getElementType();
+  if (TIndirect) {
+    unsigned AS = L->getScalarType()->getPointerAddressSpace();
+    Ty = TypedPointerType::get(Ty, AS);
+    if (auto *VT = dyn_cast<VectorType>(L))
+      Ty = VectorType::get(Ty, VT->getElementCount());
+  }
+  return Ty;
+}
+
+template <typename Fn> Type *mutateType(Type *T, Fn MutatePointer) {
+  if (T->isPointerTy()) {
+    return MutatePointer(T->getPointerAddressSpace());
+  }
+  if (auto *VT = dyn_cast<VectorType>(T)) {
+    return VectorType::get(mutateType(VT->getScalarType(), MutatePointer),
+                           VT->getElementCount());
+  }
+  if (auto *AT = dyn_cast<ArrayType>(T)) {
+    return ArrayType::get(mutateType(AT->getElementType(), MutatePointer),
+                          AT->getNumElements());
+  }
+  if (auto *FT = dyn_cast<FunctionType>(T)) {
+    SmallVector<Type *, 4> ParamTypes;
+    for (Type *Inner : FT->params())
+      ParamTypes.push_back(mutateType(Inner, MutatePointer));
+    Type *ReturnTy = mutateType(FT->getReturnType(), MutatePointer);
+    return FunctionType::get(ReturnTy, ParamTypes, FT->isVarArg());
+  }
+  // TODO: support literal structs
+  return T;
+}
+
+/// Return true if the type is an opaque pointer, or contains an opaque pointer
+/// that needs to be typed.
+bool hasPointerType(Type *T) {
+  if (T->isPtrOrPtrVectorTy())
+    return true;
+  if (auto *AT = dyn_cast<ArrayType>(T))
+    return hasPointerType(AT->getElementType());
+  if (auto *FT = dyn_cast<FunctionType>(T)) {
+    for (Type *Inner : FT->params())
+      if (hasPointerType(Inner))
+        return true;
+    return hasPointerType(FT->getReturnType());
+  }
+  // TODO: literal structs
+  return false;
+}
+
+/// Get a type where all internal pointer types are replaced with i8*.
+Type *getUnknownTyped(Type *T) {
+  Type *Int8Ty = Type::getInt8Ty(T->getContext());
+  return mutateType(
+      T, [=](unsigned AS) { return TypedPointerType::get(Int8Ty, AS); });
+}
+
+bool hasTypeVariable(Type *T, unsigned TypeVarNum) {
```

Comment:
More for readers then for the compiler, since for example isTypeVariable modifies the CheckNum which is used in the same context.
```suggestion
bool hasTypeVariable(Type *T, const unsigned TypeVarNum) {
```